pkg/metrics/metrics_test.go

```diff
@@ -315,38 +272,209 @@ func TestWriteMetric(t *testing.T) {
 			if tc.responseCodeOverride != 0 {
-				relativePath = fmt.Sprintf("%s/%d", relativePath, tc.responseCodeOverride)
+				relativePath = fmt.Sprintf("%d", tc.responseCodeOverride)
 			}
-			tc.client.Config.ServerURL = fmt.Sprintf("%s%s", ts.URL, relativePath)
+			tc.client.Config.ServerURL = fmt.Sprintf("%s/%s", ts.URL, relativePath)
 
 			err := tc.client.WriteMetric(ctx, tc.metric, tc.count)
 
 			if diff := testutil.DiffErrString(err, tc.wantErr); diff != "" {
 				t.Error(diff)
 			}
-			val, ok := reqMap.Load(relativePath + "/sendMetrics")
 			if tc.wantRequest != nil {
-				if !ok {
+				if saveReq == nil {
 					t.Errorf("no http request received, expected body of: %v", *tc.wantRequest)
 				}
-				request, ok := val.(*http.Request)
-				if !ok {
-					t.Fatal("Expected *httpRequest in sync map, but cast failed.")
+				defer saveReq.Body.Close()
+
+				var got SendMetricRequest
+				if err := json.NewDecoder(saveReq.Body).Decode(&got); err != nil {
+					t.Errorf("error reading request to test server: %s", err.Error())
+				}
+				if diff := cmp.Diff(&got, tc.wantRequest); diff != "" {
+					t.Errorf("unexpected request body. Diff (-got +want): %s", diff)
+				}
+			} else {
+				if saveReq != nil {
+					t.Errorf("did not expect a request but got one")
+				}
+			}
+		})
+	}
+}
+
+func TestWriteMetricAsync(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name          string
+		metric        string
+		count         int64
+		client        *client
+		timeoutMillis int
+		wantRequest   *SendMetricRequest
+		wantErr       string
+	}{
+		{
+			name:   "metric_success",
+			metric: "foo",
+			count:  1,
+			client: defaultClient(),
+			wantRequest: &SendMetricRequest{
+				AppID:      testAppID,
+				AppVersion: testVersion,
+				Metrics:    map[string]int64{"foo": 1},
+				InstallID:  testInstallID,
+			},
+		},
+		{
+			name:          "metric_success_timeout_set",
+			metric:        "foo",
+			count:         1,
+			client:        defaultClient(),
+			timeoutMillis: 4000,
+			wantRequest: &SendMetricRequest{
+				AppID:      testAppID,
+				AppVersion: testVersion,
+				Metrics:    map[string]int64{"foo": 1},
+				InstallID:  testInstallID,
+			},
+		},
+		{
+			name:   "metric_opt_out_noop",
+			metric: "foo",
+			count:  1,
+			client: func() *client {
+				c := defaultClient()
+				c.OptOut = true
+				return c
+			}(),
+			wantRequest: nil,
+		},
+		{
+			name:          "metric_failure_timeout",
+			metric:        "foo",
+			count:         1,
+			client:        defaultClient(),
+			timeoutMillis: 9,
+			wantRequest:   nil,
+			wantErr:       "context deadline exceeded",
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			var saveReqMutex sync.Mutex
+			var saveReq *http.Request
+
+			// Request body is intentionally leaked to allow for inspection in test cases.
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
Can this handler function be factored out since it appears in both tests? Or maybe it's substantially different between the tests, it's hard to visually diff. 

Maybe something like:

`func testHandler(w http.ResponseWriter, r *http.Request) {
    ... blah blah
}`

and in the test:

`ts := httptest.NewServer(testHandler)`

or anything approximately like that.

Reply:
Yes, I think it can. There are some slight changes between them but I think I can make it work.

Reply:
Though doing so wouldn't be quiet that clean, as I'm taking advantage of it being a closure, so I would need to make arguments to the handler explicit.

Reply:
I'm fine if you want to leave it.

Reply:
I think I will, as this was making it uglier to implement seth's suggestion of doing the validation logic within the handler function.